src/libs/OptionsListUtilsLHN.js

```diff
@@ -0,0 +1,312 @@
+import Onyx from 'react-native-onyx';
+import _ from 'underscore';
+import Str from 'expensify-common/lib/str';
+import ONYXKEYS from '../ONYXKEYS';
+import * as ReportUtils from './ReportUtils';
+import * as Localize from './Localize';
+import CONST from '../CONST';
+import * as OptionsListUtils from './OptionsListUtils';
+import * as CollectionUtils from './CollectionUtils';
+import Permissions from './Permissions';
+
+let reports;
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.REPORT,
+    waitForCollectionCallback: true,
+    callback: val => reports = val,
+});
+
+let personalDetails;
+Onyx.connect({
+    key: ONYXKEYS.PERSONAL_DETAILS,
+    callback: val => personalDetails = val,
+});
+
+let currentlyViewedReportID;
+Onyx.connect({
+    key: ONYXKEYS.CURRENTLY_VIEWED_REPORTID,
+    callback: val => currentlyViewedReportID = val,
+});
+
+let priorityMode;
+Onyx.connect({
+    key: ONYXKEYS.NVP_PRIORITY_MODE,
+    callback: val => priorityMode = val,
+});
+
+let betas;
+Onyx.connect({
+    key: ONYXKEYS.NVP_PRIORITY_MODE,
+    callback: val => betas = val,
+});
+
+const lastReportActions = {};
+const reportActions = {};
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.REPORT_ACTIONS,
+    callback: (actions, key) => {
+        if (!key || !actions) {
+            return;
+        }
+        const reportID = CollectionUtils.extractCollectionItemID(key);
+        lastReportActions[reportID] = _.last(_.toArray(actions));
+        reportActions[key] = actions;
+    },
+});
+
+let policies;
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.POLICY,
+    waitForCollectionCallback: true,
+    callback: val => policies = val,
+});
+
+let iouReports;
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.REPORT_IOUS,
+    waitForCollectionCallback: true,
+    callback: val => iouReports = val,
+});
+
+let currentUserLogin;
+Onyx.connect({
+    key: ONYXKEYS.SESSION,
+    callback: val => currentUserLogin = val,
+});
+
+let preferredLocale;
+Onyx.connect({
+    key: ONYXKEYS.NVP_PREFERRED_LOCALE,
+    callback: val => preferredLocale = val || CONST.DEFAULT_LOCALE,
+});
+
+/**
+ * @returns {String[]} An array of reportIDs sorted in the proper order
+ */
+function getOrderedReportIDs() {
+    const hideReadReports = priorityMode === CONST.PRIORITY_MODE.GSD;
+    const sortByTimestampDescending = priorityMode !== CONST.PRIORITY_MODE.GSD;
```

Comment:
Yes, there is lots of stuff duplicated now. It was helpful for me for several reasons:

1. It ensured there would be no regressions with existing components
2. It allowed me to recode everything so that I fully understood what was going on

I know you've mentioned at several points that you regret the state of the code that `OptionListUtils` ended up in. I'm curious what practical takeaways you had from that? I don't know that I really ever heard those.

Reply:
I think the core problem is that:

- `getOptions()` is a loop
- There are "specific" ways to configure `getOptions()` 
- There are groups of configs for specific use cases
- But the logic for the more specific things are added directly to `getOptions()` 
- That makes the code hard to reason about
 
And this is mostly because it's... a single loop where all the options get built.

That's kind of the fundamental problem to me since you _could_ possibly do something like:

- `getOptions(sources)` -> `getSidebarOptions(orderedOptions)`
- `createOption(optionID)` -> `createSidebarOption(option)`

But what we have instead is more like...

- `getOptions(sources, doThis, doThat, doOtherThing)`
- `createOption(id, doThis, doThat, doOtherThing)`